### ISSUE TITLE
aws-sdk-s3 upload fails with Deadlock

### ISSUE THREAD

**Issue by kplimack** [aws/aws-sdk-ruby#1832]
### Issue description
When uploading a file, a deadlock is thrown by the multipart-upload method
```
[2] pry(main)> obj.upload_file(path)
fatal: No live threads left. Deadlock?
from /opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:145:in `value'
```

### Gem name ('aws-sdk', 'aws-sdk-resources' or service gems like 'aws-sdk-s3') and its version
```
aws-sdk-core (3.20.2)
aws-sdk-kms (1.5.0)
aws-sdk-s3 (1.10.0)
```

### Version of Ruby, OS environment
```
ruby 2.1.3p242 (2014-09-19 revision 47630) [x86_64-linux-gnu]
centos 6.9
```

### Code snippets / steps to reproduce

```ruby
def upload_backup(path)
   s3 = Aws::S3::Resource.new(region: 'us-east-1')
   filename = File.basename(path)
   bucket = s3.bucket('myBackups')
   raise BackupAlreadyExistsOnS3, "file already exists in s3 #{filename}" if 
   bucket.object(filename).exists?
   obj = bucket.object(filename)
   puts "Beginning File Upload (#{filename})".light_blue
   obj.upload_file(path)
   puts "File Upload Successful (#{filename})".green
end
```

**Comment by cjyclaire** [comment]
@kplimack Quick follow-up question to help me reproduce the issue, does this happen every time every file for you or just occasionally? A rough size of the file? Is this script ran in a multithread environment?

**Comment by kplimack** [comment]
@cjyclaire this just started yesterday. We accidentally changed the IAM policy attached to this user which caused access-denied errors.  Once we fixed the IAM issue, this error cropped up.  I've tried a few different files and they all have the same issue. The files range in size from 5-5.6G. and Yes, its a multithreaded environment

**Comment by kplimack** [comment]
```ruby
[3] pry(#<Aws::S3::MultipartFileUploader>)> threads
=> [#<Thread:0x0000000806dbe8@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>,
 #<Thread:0x0000000818da50@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>,
 #<Thread:0x000000082a6248@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>,
 #<Thread:0x000000083d5538@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>,
 #<Thread:0x0000000854e1f8@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>,
 #<Thread:0x00000007ff52d8@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>,
 #<Thread:0x00000006c5e918@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>,
 #<Thread:0x00000005d130d0@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>]
[4] pry(#<Aws::S3::MultipartFileUploader>)> threads[0]
=> #<Thread:0x0000000806dbe8@/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-s3-1.12.0/lib/aws-sdk-s3/multipart_file_uploader.rb:129 sleep>
[5] pry(#<Aws::S3::MultipartFileUploader>)> threads[0].value
fatal: No live threads left. Deadlock?
from (pry):20:in `value'
```

**Comment by kplimack** [comment]
@cjyclaire I just made a 10Mb file with `dd` and it uploaded right away, this might be related to the filesize.

**Comment by kplimack** [comment]
@cjyclaire i also tried using the aws-cli and using s3-cp, it worked, too.  something with the ruby libs seems to be the issue

**Comment by peggles2** [comment]
I am having the same problem. When it gets the Aws::S3::MultipartFileUploader and creates the threads the whole thing just hangs.

**Comment by cjyclaire** [comment]
@peggles2 Are you also experiencing this in a multithread env? Or happening when you are using sdk in your APP? What's the APP environment? Does this happens all the time in your case as well?

@kplimack Soft updates that I'm still working on reproducing the issue, so far failed to reproduce on my mac osx with 5g files with 10 threads in parallel. I'll try with more thread or spin an instance with similar OS and ruby version that you mentioned : ) Let me know if you have any updates on your side that might help me reproduce or invetigate :D

**Comment by peggles2** [comment]
Hi, not sure.  How can I check if it is or not multithread env?  Its running with thin.

**Comment by srchase** [comment]
@peggles2 

Thin can be configured to run as threaded, but is not by default.

If you're using a configuration file for Thin, you could check to there: https://github.com/macournoyer/thin#configuration-files

**Comment by awood45** [comment]
We've been able to recreate this now, and are trying to nail down where this is coming from.

**Comment by awood45** [comment]
What's interesting is, I can't recreate this in the latest version of the `aws-sdk-s3` gem - looking for any relevant changes we've made.

**Comment by awood45** [comment]
Minimal reproducing example (hat tip to @srchase)

Create a file: `dd if=/dev/zero of=fifteen_mb.txt count=1024 bs=15400`

Run:

```ruby
require 'aws-sdk-s3'

s3 = Aws::S3::Resource.new(region: "us-east-1")
o = s3.bucket(BUCKET_NAME).object("mp_deadlock_test.txt")
o.upload_file("fifteen_mb.txt")
```

**Comment by awood45** [comment]
I think this was actually fixed from version 1.16.0 onwards, but please do check back and let me know if an upgrade solves your problem.

**Comment by awood45** [comment]
Tentatively putting the bug label back on because at minimum we have really inconsistent performance for the working versions as well, which could indicate the potential for these issues with bigger files.

**Comment by awood45** [comment]
Okay, we've nailed this down exactly. #1545 predicted the possibility of certain file upload situations creating an endless loop in generating checksums, and that's exactly what you've encountered. That pull request fixed this issue, and was released as `aws-sdk-s3` version `1.16.0`. If you upgrade, the problem will end.

**Comment by kplimack** [comment]
👍 thank you!
